src/main/java/com/linecorp/armeria/client/ClientInvocationHandler.java

```diff
@@ -0,0 +1,112 @@
+ * under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.UndeclaredThrowableException;
+import java.net.URI;
+import java.nio.channels.ClosedChannelException;
+import java.util.stream.Stream;
+
+import io.netty.util.concurrent.Future;
+import io.netty.util.internal.EmptyArrays;
+
+final class ClientInvocationHandler implements InvocationHandler {
+
+    private static final Method OBJECT_HASHCODE = objectMethod("hashCode");
+    private static final Method OBJECT_EQUALS = objectMethod("equals", Object.class);
+    private static final Method OBJECT_TOSTRING = objectMethod("toString");
+
+    private static Method objectMethod(String name, Class<?>... paramTypes) {
+        try {
+            return Object.class.getMethod(name, paramTypes);
+        } catch (NoSuchMethodException e) {
+            throw new Error(e); // Should never happen
+        }
+    }
+
+    private final URI uri;
+    private final Class<?> interfaceClass;
+    private final RemoteInvoker remoteInvoker;
+    private final ClientCodec codec;
+    private final ClientOptions options;
+
+    ClientInvocationHandler(URI uri, Class<?> interfaceClass,
+                            RemoteInvoker remoteInvoker, ClientCodec codec, ClientOptions options) {
+
+        this.uri = uri;
+        this.interfaceClass = interfaceClass;
+        this.remoteInvoker = remoteInvoker;
+        this.codec = codec;
+        this.options = options;
+    }
+
+    @Override
+    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+        if (method.getDeclaringClass() == Object.class) {
+            // Handle the methods in Object
+            return invokeObjectMethod(proxy, method, args);
+        } else {
+            assert method.getDeclaringClass() == interfaceClass;
+            // Handle the methods in the interface.
+            return invokeClientMethod(method, args);
+        }
+    }
+
+    private Object invokeObjectMethod(Object proxy, Method method, Object[] args) {
+        if (OBJECT_TOSTRING.equals(method)) {
+            return interfaceClass.getSimpleName() + '(' + uri + ')';
+        }
+        if (OBJECT_HASHCODE.equals(method)) {
+            return System.identityHashCode(proxy);
+        }
+        if (OBJECT_EQUALS.equals(method)) {
+            return proxy == args[0];
+        }
+
+        throw new Error("unknown method: " + method);
+    }
+
+    private Object invokeClientMethod(Method method, Object[] args) throws Throwable {
+        if (args == null) {
+            args = EmptyArrays.EMPTY_OBJECTS;
+        }
+
+        try {
+            Future<Object> resultFuture = remoteInvoker.invoke(uri, options, codec, method, args);
+            if (codec.isAsyncClient()) {
+                return method.getReturnType().isInstance(resultFuture) ? resultFuture : null;
```

Comment:
Unrelated to this PR but I've always wondered why this returns null instead of throwing an exception,  seems like programming error.